internal/engine/k8swatch/manifest_subscriber.go

```diff
@@ -0,0 +1,178 @@
+	return types.NamespacedName{Name: apis.SanitizeName(mn.String())}
+}
+
+func labelsFromSelector(selector labels.Selector) ([]v1alpha1.LabelValue, error) {
+	var out []v1alpha1.LabelValue
+	requirements, _ := selector.Requirements()
+	for _, r := range requirements {
+		if r.Operator() != selection.Equals {
+			return nil, fmt.Errorf("label %q has unsupported operator: %q", r.Key(), r.Operator())
+		}
+		values := r.Values().List()
+		if len(values) == 0 {
+			continue
+		}
+		if len(values) != 1 {
+			return nil, fmt.Errorf("label %q has more than one value: %v", r.Key(), r.Values())
+		}
+		out = append(out, v1alpha1.LabelValue{Label: r.Key(), Value: values[0]})
+	}
+	return out, nil
+}
+
+func (m *ManifestSubscriber) kubernetesDiscoveryFromManifest(ctx context.Context, key types.NamespacedName, mt *store.ManifestTarget) *v1alpha1.KubernetesDiscovery {
+	if !mt.Manifest.IsK8s() {
+		return nil
+	}
+	kt := mt.Manifest.K8sTarget()
+
+	krs := mt.State.K8sRuntimeState()
+	if len(kt.ObjectRefs) == 0 {
+		// there is nothing to discover
+		return nil
+	}
+
+	seenNamespaces := make(map[k8s.Namespace]bool)
+	var watchRefs []v1alpha1.KubernetesWatchRef
+	for _, e := range krs.DeployedEntities {
+		ns := k8s.Namespace(e.Namespace)
+		if ns == "" {
+			// since this entity is actually deployed, don't fallback to cfgNS
+			ns = k8s.DefaultNamespace
+		}
+		seenNamespaces[ns] = true
+		watchRefs = append(watchRefs, v1alpha1.KubernetesWatchRef{
+			UID:       string(e.UID),
+			Namespace: ns.String(),
+			Name:      e.Name,
+		})
+	}
+
+	for i := range kt.ObjectRefs {
+		ns := k8s.Namespace(kt.ObjectRefs[i].Namespace)
+		if ns == "" {
+			ns = m.cfgNS
+		}
+		if ns == "" {
+			ns = k8s.DefaultNamespace
+		}
+		if !seenNamespaces[ns] {
+			seenNamespaces[ns] = true
+			watchRefs = append(watchRefs, v1alpha1.KubernetesWatchRef{
+				Namespace: ns.String(),
+			})
+		}
+	}
+
+	var labelSets [][]v1alpha1.LabelValue
+	if len(seenNamespaces) > 0 {
+		for i := range kt.ExtraPodSelectors {
+			l, err := labelsFromSelector(kt.ExtraPodSelectors[i])
+			if err != nil {
+				logger.Get(ctx).Debugf("Error processing extra pod selectors for %q: %v", kt.Name, err)
+				return nil
+			}
+			labelSets = append(labelSets, l)
+		}
+	}
+
+	kd := &v1alpha1.KubernetesDiscovery{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      key.Name,
+			Namespace: key.Namespace,
+			Annotations: map[string]string{
+				v1alpha1.AnnotationManifest: mt.Manifest.Name.String(),
+				v1alpha1.AnnotationTargetID: mt.State.TargetID().String(),
+			},
+		},
+		Spec: v1alpha1.KubernetesDiscoverySpec{
+			Watches:        watchRefs,
+			ExtraSelectors: labelSets,
+		},
+	}
+
+	if fieldErrs := kd.Validate(ctx); len(fieldErrs) != 0 {
+		var sb strings.Builder
+		for i, fieldErr := range fieldErrs {
+			if i != 0 {
+				sb.WriteString(", ")
+			}
+			sb.WriteString(fieldErr.Error())
+		}
+		logger.Get(ctx).Debugf("Manifest %q produced invalid spec: %s",
```

Comment:
ditto here re: infinite loops

Reply:
This now results in an `ErrorAction` as of 3f33963be7b563808a88358a01a2a71b31c7f0b5 - comments also updated to reflect that any errors returned from `kubernetesDiscoveryFromManifest` are expected to be fatal and some explanation re: why those errors are indicative of a Tilt bug and should be impossible unless we regress something.